internal/xds/rbac/rbac_engine_test.go

```diff
@@ -1007,85 +1224,704 @@ func (s) TestChainEngine(t *testing.T) {
+					}
+					if !reflect.DeepEqual(b.auditEvents, data.wantAuditEvents) {
+						t.Fatalf("Unexpected audit event for query:%v", data)
+					}
+
+					// this builder's auditEvents can be shared for several queries, make sure it's empty
+					b.auditEvents = nil
+				}()
+			}
+		})
+	}
+}
+
+type ServerTransportStreamWithMethod struct {
+	method string
+}
+
+func (sts *ServerTransportStreamWithMethod) Method() string {
+	return sts.method
+}
+
+func (sts *ServerTransportStreamWithMethod) SetHeader(md metadata.MD) error {
+	return nil
+}
+
+func (sts *ServerTransportStreamWithMethod) SendHeader(md metadata.MD) error {
+	return nil
+}
+
+func (sts *ServerTransportStreamWithMethod) SetTrailer(md metadata.MD) error {
+	return nil
+}
+
+// An audit logger that will log to the auditEvents slice
+type TestAuditLoggerBuffer struct {
+	auditEvents *[]*audit.Event
+}
+
+func (logger *TestAuditLoggerBuffer) Log(e *audit.Event) {
+	*(logger.auditEvents) = append(*(logger.auditEvents), e)
+}
+
+// Builds TestAuditLoggerBuffer
+type TestAuditLoggerBufferBuilder struct {
+	auditEvents []*audit.Event
+}
+
+// The required config for TestAuditLoggerBuffer
+type TestAuditLoggerBufferConfig struct {
+	audit.LoggerConfig
+}
+
+func (builder *TestAuditLoggerBufferBuilder) ParseLoggerConfig(configJSON json.RawMessage) (config audit.LoggerConfig, err error) {
+	return TestAuditLoggerBufferConfig{}, nil
+}
+
+func (builder *TestAuditLoggerBufferBuilder) Build(config audit.LoggerConfig) audit.Logger {
+	return &TestAuditLoggerBuffer{auditEvents: &builder.auditEvents}
+}
+
+func (builder *TestAuditLoggerBufferBuilder) Name() string {
+	return "TestAuditLoggerBuffer"
+}
+
+// An audit logger to test using a custom config
+type TestAuditLoggerCustomConfig struct{}
+
+func (logger *TestAuditLoggerCustomConfig) Log(*audit.Event) {}
+
+// Build TestAuditLoggerCustomConfig
+type TestAuditLoggerCustomConfigBuilder struct {
+}
+
+// The custom config for the TestAuditLoggerCustomConfig logger
+type TestAuditLoggerCustomConfigConfig struct {
+	audit.LoggerConfig
+	Abc int
+	Xyz string
+}
+
+// Parses TestAuditLoggerCustomConfigConfig
+// Hard-coded to match with it's test case above
+func (builder TestAuditLoggerCustomConfigBuilder) ParseLoggerConfig(configJSON json.RawMessage) (audit.LoggerConfig, error) {
+	c := TestAuditLoggerCustomConfigConfig{}
+	err := json.Unmarshal(configJSON, &c)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse custom config: %v", err)
+	}
+	return c, nil
+}
+
+func (builder *TestAuditLoggerCustomConfigBuilder) Build(config audit.LoggerConfig) audit.Logger {
+	return &TestAuditLoggerCustomConfig{}
+}
+
+func (builder *TestAuditLoggerCustomConfigBuilder) Name() string {
+	return "TestAuditLoggerCustomConfig"
+}
+
+// Builds custom configs for audit logger RBAC protos
+func createUDPATypedStruct(t *testing.T, in map[string]interface{}, name string) *anypb.Any {
+	t.Helper()
```

Comment:
`createUDPATypedStruct` does qualify to be a t.Helper(). But I'm concerned about calling `t.Fatal` in here. 

In test when this helper errors out, marking this method as t.Helper would make it harder to find the issue from stack trace. Especially since there are a couple of fatal paths below. 

I would recommend either just nixing `t.Helper` from this method and also `createXDSTypedStruct` below.

Reply:
I added the `t.Helper` because @easwars recommended it in a similar case in another PR
I'll do option 1 - the main test uses this in a huge struct literal so I don't want to have to handle passing the errors back out.

So doing option 1 should I remove `t.Helper`?

Reply:
See go/go-style/decisions#mark-test-helpers for more info on test helpers.

Reply:
Added more detail